File: src/Cyborg/Core/Area.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Cyborg.Components;
using Cyborg.ContentPipeline.Maps;
using Cyborg.Entities;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Content;
using Microsoft.Xna.Framework.Graphics;

namespace Cyborg.Core
{
    public class Area : IArea
    {
        private readonly IEntityManager _entityManager;
        private readonly string _mapName;

        private IEnumerable<IEntity> _loadedEntites = null;

        public Area(IEntityManager entityManager, string mapName, Rectangle areaBounds)
        {
            _entityManager = entityManager;
            _mapName = mapName;

            Bounds = areaBounds;
        }

        public Rectangle Bounds { get; }

        public void Load()
        {
            if (_loadedEntites != null)
                return;

            _loadedEntites = _entityManager.CreateMany(cm => CreateAreaEntities(cm));
        }

        public void Unload()
        {
            if (_loadedEntites == null)
                return;

            foreach (var entity in _loadedEntites)
                entity.Destroyed = true;

            _loadedEntites = null;
        }

        private IEnumerable<IEntity> CreateAreaEntities(ContentManager contentManager)
        {
            var map = contentManager.Load<Map>($"Maps/{_mapName}");
            var spriteSheet = contentManager.Load<Texture2D>($"Maps/{map.TileSet.SpriteSheet}");

            // Passthrough Underlay
            var passThroughUnderlay = map.Tiles["floor"];
            for (var x = 0; x < passThroughUnderlay.GetLength(0); x++)
                for (var y = 0; y < passThroughUnderlay.GetLength(1); y++)
                    if (passThroughUnderlay[x, y] > 0)
                    {
                        var tileBounds = GetTileBounds(x, y);
                        if (!tileBounds.Intersects(Bounds))
                            continue;

                        var spriteFrame = GetTileSpriteFrame(passThroughUnderlay[x, y]);
                        yield return new Tile(tileBounds.Center.ToVector2(), tileBounds.Size, Edge.None, spriteSheet, spriteFrame, 1);
                    }

            // Obstacle Underlay
            var obstacleUnderlay = map.Tiles["walls"];
            for (var x = 0; x < obstacleUnderlay.GetLength(0); x++)
                for (var y = 0; y < obstacleUnderlay.GetLength(1); y++)
                    if (obstacleUnderlay[x, y] > 0)
                    {
                        var tileBounds = GetTileBounds(x, y);
                        if (!tileBounds.Intersects(Bounds))
                            continue;

                        var edges = Edge.None;
                        if (x > 0 && obstacleUnderlay[x - 1, y] == 0)
                            edges |= Edge.Left;
                        if (x < obstacleUnderlay.GetLength(0) - 1 && obstacleUnderlay[x + 1, y] == 0)
                            edges |= Edge.Right;
                        if (y > 0 && obstacleUnderlay[x, y - 1] == 0)
                            edges |= Edge.Top;
                        if (y < obstacleUnderlay.GetLength(1) - 1 && obstacleUnderlay[x, y + 1] == 0)
                            edges |= Edge.Bottom;
                        var spriteFrame = GetTileSpriteFrame(obstacleUnderlay[x, y]);
                        yield return new Tile(tileBounds.Center.ToVector2(), tileBounds.Size, edges, spriteSheet, spriteFrame, 2);
                    }

            // Passthrough Overlay
            var passthroughOverlay = map.Tiles["overlay"];
            for (var x = 0; x < passthroughOverlay.GetLength(0); x++)
                for (var y = 0; y < passthroughOverlay.GetLength(1); y++)
                    if (passthroughOverlay[x, y] > 0)
                    {
                        var tileBounds = GetTileBounds(x, y);
                        if (!tileBounds.Intersects(Bounds))
                            continue;

                        var spriteFrame = GetTileSpriteFrame(passthroughOverlay[x, y]);
                        yield return new Tile(tileBounds.Center.ToVector2(), tileBounds.Size, Edge.None, spriteSheet, spriteFrame, 5);
                    }

            // Enemies
            foreach (var enemy in map.Objects["enemies"])
            {
                var enemyBounds = new Rectangle(enemy.X, enemy.Y, enemy.Width, enemy.Height);
                if (!enemyBounds.Intersects(Bounds))
                    continue;

                if (enemy.Type == "Zombie")
                    yield return new Zombie(contentManager, enemyBounds.Center.ToVector2());
            }

            Rectangle GetTileBounds(int xIndex, int yIndex) => new(xIndex * map.TileSet.TileWidth, yIndex * map.TileSet.TileHeight, map.TileSet.TileWidth, map.TileSet.TileHeight);

            Rectangle GetTileSpriteFrame(int value)
            {
                var tileIndex = value - 1;
                var spriteFrameOffsetX = tileIndex % map.TileSet.Columns * map.TileSet.TileWidth;
                var spriteFrameOffsetY = tileIndex / map.TileSet.Columns * map.TileSet.TileHeight;
                return new(spriteFrameOffsetX, spriteFrameOffsetY, map.TileSet.TileWidth, map.TileSet.TileHeight);
            }
        }
    }
}

File: src/systems/area.ts
import { System } from ".";
import * as demoMap from "../assets/maps/demo_map.json";
import { BaseComponent, Edges, getBounds, hasBody } from "../components";
import { isPlayer } from "../components";
import { Tile, createTile } from "../entities";
import { createZombie } from "../entities/zombie";
import { camera } from "../framework/camera";
import { gameState } from "../framework/gameState";
import { getResource, Resource } from "../framework/resources";
import { Rectangle, intersects, centre, liesWithin } from "../shapes";

const areaLayer = demoMap.layers.filter((l) => l.name === "areas")[0];
if (!areaLayer || !areaLayer.objects) throw new Error("Invalid layer provided.");

const areas: ReadonlyArray<Rectangle> = areaLayer.objects;
const areaEntites: BaseComponent[][] = areas.map((_) => []);

const transitionDuration = 0.5;
let currentAreaIndex: number | undefined;
let previousAreaIndex: number | undefined;
let transitionElapsed: number = 0;

const areaSystem: System = (entities, _, deltaSeconds) => {
	checkAreaTransition(entities);
	applyAreaTransition(deltaSeconds);
};

const checkAreaTransition = (entities: BaseComponent[]): void => {
	for (const entity of entities) {
		if (!isPlayer(entity) || !hasBody(entity)) continue;

		const playerBounds = getBounds(entity);
		const newAreaIndex = areas.findIndex((area) => intersects(area, playerBounds));
		if (newAreaIndex < 0) break;

		// Check begin transition
		if (!gameState.transitioning && newAreaIndex !== currentAreaIndex) {
			gameState.transitioning = true;

			previousAreaIndex = currentAreaIndex;
			currentAreaIndex = newAreaIndex;
			transitionElapsed = 0;

			loadArea(entities, currentAreaIndex);
		}

		// Check end transition
		if (gameState.transitioning && transitionElapsed > transitionDuration && typeof currentAreaIndex !== "undefined") {
			gameState.transitioning = false;

			camera.x = areas[currentAreaIndex].x;
			camera.y = areas[currentAreaIndex].y;

			if (typeof previousAreaIndex !== "undefined") unloadArea(previousAreaIndex);
		}
	}
};

const applyAreaTransition = (deltaSeconds: number): void => {
	if (!gameState.transitioning || typeof currentAreaIndex === "undefined") return;

	transitionElapsed += deltaSeconds;

	const progress = transitionElapsed / transitionDuration;
	const startArea = areas[typeof previousAreaIndex !== "undefined" ? previousAreaIndex : currentAreaIndex];
	const endArea = areas[currentAreaIndex];

	camera.x = startArea.x + (endArea.x - startArea.x) * progress;
	camera.y = startArea.y + (endArea.y - startArea.y) * progress;
};

const loadArea = (entities: BaseComponent[], areaIndex: number): void => {
	const area = areas[areaIndex];
	const floorTiles = createTiles(area, "floor", false, 0);
	const wallTiles = createTiles(area, "walls", true, 1);
	const overlayTiles = createTiles(area, "overlay", false, 3);
	const enemies = createEnemies(area);

	const newEntities = [...floorTiles, ...wallTiles, ...overlayTiles, ...enemies];
	areaEntites[areaIndex] = newEntities;
	entities.push(...newEntities);
};

const createTiles = (area: Rectangle, layer: string, solid: boolean, zIndex: number): Tile[] => {
	const tilesLayer = demoMap.layers.filter((l) => l.name === layer)[0];
	const tileSet = demoMap.tilesets[0];
	const textureAtlas = getResource(Resource.DemoMap).texture;

	if (!tilesLayer || tilesLayer.type !== "tilelayer" || !tilesLayer.data || !tilesLayer.width || !tilesLayer.height)
		throw new Error("Invalid layer provided.");

	const tiles: Tile[] = [];
	for (let yIndex = 0; yIndex < tilesLayer.height; yIndex++)
		for (let xIndex = 0; xIndex < tilesLayer.width; xIndex++) {
			const tileBounds: Rectangle = {
				x: xIndex * demoMap.tilewidth,
				y: yIndex * demoMap.tileheight,
				width: demoMap.tilewidth,
				height: demoMap.tileheight
			};
			if (!intersects(area, tileBounds)) continue;

			const index = xIndex + yIndex * tilesLayer.width;
			const tileValue = tilesLayer.data[index];
			if (!tileValue) continue;

			const textureAtlasIndex = tileValue - 1;
			const textureAtlasX = (textureAtlasIndex % tileSet.columns) * tileSet.tilewidth;
			const textureAtlasY = Math.trunc(textureAtlasIndex / tileSet.columns) * tileSet.tileheight;
			const textureAtlasFrame: Rectangle = {
				x: textureAtlasX,
				y: textureAtlasY,
				width: tileSet.tilewidth,
				height: tileSet.tileheight
			};

			const edges: Edges = { bottom: true, left: true, right: true, top: true };
			if (!solid || (yIndex < tilesLayer.height - 1 && tilesLayer.data[xIndex + (yIndex + 1) * tilesLayer.width])) edges.bottom = false;
			if (!solid || (xIndex > 0 && tilesLayer.data[xIndex - 1 + yIndex * tilesLayer.width])) edges.left = false;
			if (!solid || (xIndex < tilesLayer.width - 1 && tilesLayer.data[xIndex + 1 + yIndex * tilesLayer.width])) edges.right = false;
			if (!solid || (yIndex > 0 && tilesLayer.data[xIndex + (yIndex - 1) * tilesLayer.width])) edges.top = false;

			const tile: Tile = createTile(textureAtlas, textureAtlasFrame, tileBounds, edges, zIndex);
			tiles.push(tile);
		}

	return tiles;
};

const createEnemies = (area: Rectangle): BaseComponent[] => {
	const enemiesLayer = demoMap.layers.filter((l) => l.name === "enemies")[0];

	if (!enemiesLayer || enemiesLayer.type !== "objectgroup" || !enemiesLayer.objects) throw new Error("Invalid layer provided.");

	const enemies: BaseComponent[] = [];
	for (const object of enemiesLayer.objects) {
		const objectCentre = centre(object);
		if (!liesWithin(objectCentre, area)) continue;

		if (object.type === "Zombie") {
			const zombie = createZombie(centre(object));
			enemies.push(zombie);
		}
	}

	return enemies;
};

const unloadArea = (areaIndex: number): void => {
	areaEntites[areaIndex].forEach((entity) => (entity.destroyed = true));
	areaEntites[areaIndex] = [];
};

export { areaSystem };
